### ISSUE TITLE
Backspace in shell fails on some browsers/OS

### ISSUE THREAD

**Issue by mchitre** [org-arl/fjage#166]
@ngyewch was able to reproduce this on Linux
Also reported in https://stackoverflow.com/questions/65834755/how-to-use-backspace-in-unetstack-shell

**Comment by notthetup** [comment]
I tested this on Firefox on macOS. I can't seem to reproduce this. @ngyewch is this sporadic or is this something that happens all the time?

**Comment by notthetup** [comment]
Jline seems to have a bunch of updates since we picked it up.

https://github.com/jline/jline3/blob/master/changelog.md

Nothing on `xterm.js`

**Comment by ngyewch** [comment]
This happens pretty consistently on my laptop, didn't see it on my desktop:
* OS: Ubuntu 20.04
* Chrome: Version 87.0.4280.141 (Official Build) (64-bit)
* infocmp: [infocmp.txt](https://github.com/org-arl/fjage/files/5854821/infocmp.txt)

https://user-images.githubusercontent.com/3282336/105472438-e6363100-5cd6-11eb-8821-637686472dc2.mp4

**Comment by Priyansh25** [comment]
I have Ubuntu 18.04.4 LTS , I have tested on both \
Chrome : Version 86.0.4240.183 (Official Build) (64-bit) and \
firefox : 84.0.2 (64-bit) 
and the problem is same on both the browsers as told by @ngyewch.

**Comment by notthetup** [comment]
I tested on Parallels on Ubuntu 16.04 with Firefox and I am still unable to reproduce it. I will try to find a native Linux box to run it on Monday.

**Comment by manuignatius** [comment]
Tested on both Ubuntu 18.04.05 and 20.04 LTS version (running on Parallels 16) on both Chrome and Firefox, reproducible always. See below. I have seen this issue before but had assumed it was something to do with the terminal setting.


https://user-images.githubusercontent.com/4144181/105583035-f9c2c400-5dca-11eb-9b0d-109b885797c7.mov

**Comment by notthetup** [comment]
OK. I tested this on a new Ubuntu 20.04 (Parallels). And I can reproduce it using Firefox and also when I connected to the Web UI from Chrome on macOS.

So I guess it's not so much related to the front end of things (xterm.js) as much as jline.

**Comment by notthetup** [comment]
For comparison, I've two dev consoles to two different instances of fjage. One running on Ubuntu 20.04 and one locally on macOS (right).

I typed `ps` and then `Backspace`.

The fjage instance on macOS replies with the keycode for backspace `**[K` while the one on Ubuntu doesn't.. 

![Screen Shot 2021-01-25 at 17 28 25](https://user-images.githubusercontent.com/553140/105686912-c8fd9e80-5f32-11eb-826f-d0f564d5ddd3.jpg)

**Comment by notthetup** [comment]
Moving up to jline-3.19.0 doesn't seem to fix it. :(

**Comment by mchitre** [comment]
Probably to do with what terminal emulation xterm negotiates with jline.

**Comment by notthetup** [comment]
Yup. I traced through the code. It effectively does `infocmp xterm-noapp` 

Running that on a cli, seems to return valid output on the OS where the issue doesn't exist and returns an empty output where it does.

JLine's Wiki on Terminals doesn't mention `xterm-noapp`, but it has `xterm`. https://github.com/jline/jline3/wiki/Terminals

@mchitre Do you know what the `noapp` suffix is supposed to change?

**Comment by notthetup** [comment]
On macOS

```
infocmp -x xterm xterm-noapp
comparing xterm to xterm-noapp.
    comparing booleans.
    comparing numbers.
    comparing strings.
	kcub1: '\EOD', '\E[D'.
	kcud1: '\EOB', '\E[B'.
	kcuf1: '\EOC', '\E[C'.
	kcuu1: '\EOA', '\E[A'.
	kend: '\EOF', '\E[F'.
	khome: '\EOH', '\E[H'.
	rmcup: '\E[?1049l\E[23;0;0t', NULL.
	rmkx: '\E[?1l\E>', '\E>'.
	smcup: '\E[?1049h\E[22;0;0t', NULL.
	smkx: '\E[?1h\E=', '\E='.
```

**Comment by mchitre** [comment]
Related info:
https://invisible-island.net/xterm/terminfo.html

`app`/`noapp` is supposed to indicate whether the cursor keys are in "application mode" or not.

**Comment by notthetup** [comment]
> VTxxx terminals are usually set up so that full-screen applications will use
> the cursor application mode strings.  This is good for full-screen
> applications, including legacy applications which may have hard-coded
> behavior, but bad for interactive shells (e.g., tcsh, bash) which use arrow
> keys to scroll through a history of command strings.

**Comment by notthetup** [comment]
Seems like modern Ubuntu's use a `xterm-256color` as a default

**Comment by notthetup** [comment]
I tested with both `xterm-256color` and `xterm`.

With `xterm-256color` the issue seems to exist. But if I change it to `xterm`. It seems to go away.

**Comment by mchitre** [comment]
So can we force the term to `xterm` to fix this?

**Comment by notthetup** [comment]
It seems to go away when I tested that. I can build a fjage.jar upload it here. @ngyewch could you try and let me know if that fixes it for you?

I wonder though how we can test for regression?

**Comment by notthetup** [comment]
Created a branch with my change : https://github.com/org-arl/fjage/tree/term-backspace-fix

Here's a jar built on that [fjage-1.9.0-SNAPSHOT.jar.zip](https://github.com/org-arl/fjage/files/5872280/fjage-1.9.0-SNAPSHOT.jar.zip)

**Comment by ngyewch** [comment]
The fix looks good (so far). Today I tested on my Ubuntu 20.04 desktop. I saw the problem on both web (Chrome) and telnet before applying the fix.

Screenshot (telnet, before fix)
![telnet before fix 2021-01-27 15-57-36](https://user-images.githubusercontent.com/3282336/105961377-64208080-60b9-11eb-99c8-b52086bbc24a.png)

Screenshot (telnet, after fix)
![telnet after fix 2021-01-27 16-00-46](https://user-images.githubusercontent.com/3282336/105961577-a3e76800-60b9-11eb-85ce-d18b56229ca2.png)

Screenshot (web (Chrome), after fix)
![web after fix 2021-01-27 16-07-51](https://user-images.githubusercontent.com/3282336/105961775-ec9f2100-60b9-11eb-9da3-860abc142cb3.png)

**Comment by Priyansh25** [comment]
It worked on my system as well for both chrome and firefox. Thanks for the fix.

**Comment by notthetup** [comment]
OK. We need to test this on 

### Older Ubuntu <= 20.04
- [ ] Console Shell
- [ ] Web Shell (Firefox, Chrome)
- [ ] Telnet

### Newer Ubuntu 20.04 >=
- [ ] Console Shell
- [ ] Web Shell (Firefox, Chrome)
- [ ] Telnet

### macOS (Catalina/Big Sur)
- [ ] Console Shell
- [ ] Web Shell (Firefox, Chrome, Safari)
- [ ] Telnet

### Windows 10
- [ ] Console Shell
- [ ] Web Shell (Firefox, Chrome, Edge)
- [ ] Telnet

**Comment by patel999jay** [comment]
I tried on windows 10 on Web Shell(Chrome, Edge), it works fine with them.
![image](https://user-images.githubusercontent.com/5512610/106312017-6ced8f80-623c-11eb-8d9e-55f36a8f490d.png)

**Comment by notthetup** [comment]
Thanks @patel999jay !

**Comment by notthetup** [comment]
OK. I tested on all the OSes. The fix seems to work. Will PR the branch.

**Comment by notthetup** [comment]
Fixed in #176
